src/upload.rs

```diff
@@ -70,3 +94,107 @@ pub async fn upload_package_to_quetz(
     Ok(())
 }
+
+pub async fn upload_package_to_prefix(
+    storage: &AuthenticationStorage,
+    api_key: Option<String>,
+    package_file: PathBuf,
+    url: Url,
+    channel: String,
+) -> miette::Result<()> {
+    let token = match api_key {
+        Some(api_key) => api_key,
+        None => match storage.get_by_url(url.clone()) {
+            Ok((_, Some(Authentication::BearerToken(token)))) => token,
+            Ok((_, Some(_))) => {
+                return Err(miette::miette!("A Conda token is required for authentication with prefix.dev.
+                        Authentication information found in the keychain / auth file, but it was not a Bearer token"));
+            }
+            Ok((_, None)) => {
+                return Err(miette::miette!(
+                    "No prefix.dev api key was given and none was found in the keychain / auth file"
+                ));
+            }
+            Err(e) => {
+                return Err(miette::miette!(
+                    "Failed to get authentication information form keychain: {e}"
+                ));
+            }
+        },
+    };
+
+    let filename = package_file
+        .file_name()
+        .unwrap()
+        .to_string_lossy()
+        .to_string();
+
+    let file_size = package_file.metadata().into_diagnostic()?.len();
+
+    println!("Uploading package to: {}", url);
+    println!(
+        "Package file:         {} ({})\n",
+        package_file.display(),
+        HumanBytes(file_size)
+    );
+
+    let url = url
+        .join(&format!("api/v1/upload/{}", channel))
+        .into_diagnostic()?;
+
+    let client = reqwest::Client::builder()
+        .no_gzip()
+        .build()
+        .expect("failed to create client");
+
+    let sha256sum = format!(
+        "{:x}",
+        compute_file_digest::<Sha256>(&package_file).into_diagnostic()?
+    );
+
+    let file = tokio::fs::File::open(&package_file)
+        .await
+        .into_diagnostic()?;
+
+    let progress_bar = indicatif::ProgressBar::new(file_size)
+        .with_prefix("Uploading")
+        .with_style(default_bytes_style().into_diagnostic()?);
+
+    let reader_stream = ReaderStream::new(file)
+        .inspect_ok(move |bytes| {
+            progress_bar.inc(bytes.len() as u64);
+        })
+        .inspect_err(|e| {
+            println!("Error while uploading: {}", e);
+        });
+
+    let body = reqwest::Body::wrap_stream(reader_stream);
+
+    let response = client
+        .post(url.clone())
+        .header("X-File-Sha256", sha256sum)
+        .header("X-File-Name", filename)
+        .header("Content-Length", file_size)
+        .header("Content-Type", "application/octet-stream")
+        .bearer_auth(token)
+        .body(body)
+        .send()
+        .await
+        .into_diagnostic()?;
+
+    if response.status().is_success() {
+        println!("Upload successful!");
+    } else {
+        println!("Upload failed!");
+        if response.status() == 401 {
```

Comment:
You can use https://docs.rs/reqwest/latest/reqwest/struct.Response.html#method.error_for_status